src/rsp.c

```diff
@@ -56,93 +21,115 @@ static void __SP_DMA_wait(void)
     while (SP_regs->status & (SP_STATUS_DMA_BUSY | SP_STATUS_IO_BUSY)) ;
 }
 
-void rsp_init()
+void rsp_init(void)
 {
     /* Make sure RSP is halted */
-    *(volatile uint32_t *) 0xA4080000 = SP_DMA_IMEM;
-    SP_regs->status = SP_STATUS_SET_HALT;
+    *SP_PC = 0x1000;
+    SP_regs->status = SP_WSTATUS_SET_HALT;
+}
 
-    return;
+void rsp_load(rsp_ucode_t *ucode) {
+    if (cur_ucode != ucode) {
+        rsp_load_code(ucode->code, (uint8_t*)ucode->code_end - ucode->code, 0);
+        rsp_load_data(ucode->data, (uint8_t*)ucode->data_end - ucode->data, 0);
+        cur_ucode = ucode;
+    }
 }
 
-void load_ucode(void* start, unsigned long size)
+void rsp_load_code(void* start, unsigned long size, unsigned int imem_offset)
 {
+    assert(((uint32_t)start % 8) == 0);
+    assert((imem_offset % 8) == 0);
+
     disable_interrupts();
     __SP_DMA_wait();
 
     SP_regs->DRAM_addr = start;
     MEMORY_BARRIER();
-    SP_regs->RSP_addr = (void*)SP_DMA_IMEM;
+    SP_regs->RSP_addr = SP_IMEM + imem_offset/4;
     MEMORY_BARRIER();
     SP_regs->rsp_read_length = size - 1;
     MEMORY_BARRIER();
 
     __SP_DMA_wait();
     enable_interrupts();
-    return;
 }
 
-void load_data(void* start, unsigned long size)
+void rsp_load_data(void* start, unsigned long size, unsigned int dmem_offset)
 {
+    assert(((uint32_t)start % 8) == 0);
+    assert((dmem_offset % 8) == 0);
+
     disable_interrupts();
     __SP_DMA_wait();
 
     SP_regs->DRAM_addr = start;
     MEMORY_BARRIER();
-    SP_regs->RSP_addr = (void*)SP_DMA_DMEM;
+    SP_regs->RSP_addr = SP_DMEM + dmem_offset/4;
```

Comment:
Since `dmem_offset` is unsigned, would there be any rounding problems using `>> 2` instead of integer division here?

This might come across as a nitpick, but my rationale is that the extra cycles are given back to the person doing their homebrew project. Thoughts?

Reply:
Same goes for `imem_offset` in `rsp_load_code` too.

Reply:
Actually the division does not translate to a division opcode because it's added to a `uint32_t*` pointer, so the value in `dmem_offset` is directly added to the constant behind `SP_DMEM` (as you'd expect, since `dmem_offset` is a byte offset anyway).

Even if that wasn't the case, GCC is able to optimize `/4` into `>>2` on unsigned integers, just like it can optimize `%2` into `&1`.